README.md
# graphviz-jpgd
This is a clone of the original alexmerz-graphviz-jpgd lib, which was created more than 10 years ago and doesn't have a repo in github.
Visit http://www.alexander-merz.com/graphviz/ to get info for the original project.

CHANGELOG
- fix handling of generic node attributes
0.8.6beta
- change toString() implementation, now you can use
  the output to generate a graphviz document by calling 
  Graph.toString()
- fix a bug with not setting the direction type
  in edges
0.8.5beta
- add small manual
- fix handling of generic graph attributes
0.8.4beta
- add documentation in several files
- jar contains binaries only, use tgz for distribution


src/test/java/com/alexmerz/graphviz/unittest/Parser2Test.java

/***************************************************************
 *    JPGD - Java-based Parser for Graphviz Documents
 *    Copyright (C) 2006  Alexander Merz
 *
 *   This library is free software; you can redistribute it and/or
 *   modify it under the terms of the GNU Lesser General Public
 *   License as published by the Free Software Foundation; either
 *   version 2.1 of the License.
 *
 *   This library is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *   Lesser General Public License for more details.
 *
 *   You should have received a copy of the GNU Lesser General Public
 *   License along with this library; if not, write to the Free Software
 *   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */


package com.alexmerz.graphviz.unittest;

import java.io.*;
import java.util.ArrayList;

import com.alexmerz.graphviz.objects.Graph;
import junit.framework.TestCase;

import com.alexmerz.graphviz.ParseException;
import com.alexmerz.graphviz.Parser;

/**
 * This test suite uses the Graphviz examples
 * from the Gallery on the Graphviz webpage.
 * (An exception is only NodeEdge) 
 * All files are stored in the testfiles directories.
 * 
 * @version $Id: Parser2Test.java,v 1.4 2006/03/20 16:47:49 Alexander Exp $
 * @author Alexander
 *
 */
public class Parser2Test extends TestCase {

	public void testNodeEdge() throws ParseException {
        Reader in=null;
        try {
            InputStream inputStream = this.getClass().getClassLoader()
                    .getResourceAsStream("testfiles/NodeEdge.viz");
            in = new InputStreamReader(inputStream);
            Parser p = new Parser(in);            
            assertTrue(p.parse(in));  
            
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }               
	}
	
	public void testCluster() throws ParseException {
        Reader in=null;
        try {
            InputStream inputStream = this.getClass().getClassLoader()
                    .getResourceAsStream("testfiles/cluster.viz");
            in = new InputStreamReader(inputStream);
            Parser p = new Parser(in);
            assertTrue(p.parse(in));
            
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }
	}   
    
    public void testCrazy() throws ParseException {
        Reader in=null;
        try {
          	InputStream inputStream = this.getClass().getClassLoader()
                    .getResourceAsStream("testfiles/crazy.viz");
          	in = new InputStreamReader(inputStream);
            Parser p = new Parser(in);
            assertTrue(p.parse(in));                
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }        
	}
    
    public void testDatastruct() throws ParseException {
        Reader in=null;
        try {
            InputStream inputStream = this.getClass().getClassLoader()
                    .getResourceAsStream("testfiles/datastruct.viz");
            in = new InputStreamReader(inputStream);
            Parser p = new Parser(in);
            assertTrue(p.parse(in));                
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }        
	}
    
    public void testFdpclust() throws ParseException {
        Reader in=null;
        try {
            InputStream inputStream = this.getClass().getClassLoader()
                    .getResourceAsStream("testfiles/fdpclust.viz");
            in = new InputStreamReader(inputStream);
            Parser p = new Parser(in);
            assertTrue(p.parse(in));                
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }        
	}    

    public void testFSM() throws ParseException {
        Reader in=null;
        try {
            InputStream inputStream = this.getClass().getClassLoader()
                    .getResourceAsStream("testfiles/fsm.viz");
            in = new InputStreamReader(inputStream);
            Parser p = new Parser(in);
            assertTrue(p.parse(in));                
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }        
	}

    public void testHello() throws ParseException {
        Reader in=null;
        try {
            InputStream inputStream = this.getClass().getClassLoader()
                    .getResourceAsStream("testfiles/hello.viz");
            in = new InputStreamReader(inputStream);
            Parser p = new Parser(in);
            assertTrue(p.parse(in));                
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }        
	}
   
    public void testInet() throws ParseException {
        Reader in=null;
        try {
            InputStream inputStream = this.getClass().getClassLoader()
                    .getResourceAsStream("testfiles/inet.viz");
            in = new InputStreamReader(inputStream);
            Parser p = new Parser(in);
            assertTrue(p.parse(in));                
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }        
	}

    
       
    public void testProcess() throws ParseException {
        Reader in=null;
        try {
            InputStream inputStream = this.getClass().getClassLoader()
                    .getResourceAsStream("testfiles/process.viz");
            in = new InputStreamReader(inputStream);
            Parser p = new Parser(in);
            assertTrue(p.parse(in));                
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }        
	}

    public void testProfile() throws ParseException {
        Reader in=null;
        try {
            InputStream inputStream = this.getClass().getClassLoader()
                    .getResourceAsStream("testfiles/profile.viz");
            in = new InputStreamReader(inputStream);
            Parser p = new Parser(in);
            assertTrue(p.parse(in));                
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }        
	}

    public void testSdh() throws ParseException {
        Reader in=null;
        try {
            InputStream inputStream = this.getClass().getClassLoader()
                    .getResourceAsStream("testfiles/sdh.viz");
            in = new InputStreamReader(inputStream);
            Parser p = new Parser(in);
            assertTrue(p.parse(in));                
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }        
	}


    
    public void testSoftmain() throws ParseException {
        Reader in=null;
        try {
            InputStream inputStream = this.getClass().getClassLoader()
                    .getResourceAsStream("testfiles/softmain.viz");
            in = new InputStreamReader(inputStream);
            Parser p = new Parser(in);
            assertTrue(p.parse(in));                
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }        
	}

    public void testSwitch() throws ParseException {
        Reader in=null;
        try {
            InputStream inputStream = this.getClass().getClassLoader()
                    .getResourceAsStream("testfiles/switch.viz");
            in = new InputStreamReader(inputStream);
            Parser p = new Parser(in);
            assertTrue(p.parse(in));                
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }        
	}    

    
    public void testTransparency() throws ParseException {
        Reader in=null;
        try {
            InputStream inputStream = this.getClass().getClassLoader()
                    .getResourceAsStream("testfiles/transparency.viz");
            in = new InputStreamReader(inputStream);
            Parser p = new Parser(in);
            assertTrue(p.parse(in));                
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }        
	}

    
    public void testUnix() throws ParseException {
        Reader in=null;
        try {
            InputStream inputStream = this.getClass().getClassLoader()
                    .getResourceAsStream("testfiles/unix.viz");
            in = new InputStreamReader(inputStream);
            Parser p = new Parser(in);
            assertTrue(p.parse(in));                
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }        
	}    
    
    public void testWorld() throws ParseException {
        Reader in=null;
        try {
            InputStream inputStream = this.getClass().getClassLoader()
                    .getResourceAsStream("testfiles/world.viz");
            in = new InputStreamReader(inputStream);
            Parser p = new Parser(in);
            assertTrue(p.parse(in));                
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }        
	}

}


src/test/java/com/alexmerz/graphviz/unittest/Parser3Test.java
/***************************************************************
 *    JPGD - Java-based Parser for Graphviz Documents
 *    Copyright (C) 2006  Alexander Merz
 *
 *   This library is free software; you can redistribute it and/or
 *   modify it under the terms of the GNU Lesser General Public
 *   License as published by the Free Software Foundation; either
 *   version 2.1 of the License.
 *
 *   This library is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *   Lesser General Public License for more details.
 *
 *   You should have received a copy of the GNU Lesser General Public
 *   License along with this library; if not, write to the Free Software
 *   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

package com.alexmerz.graphviz.unittest;

import java.io.*;
import java.util.ArrayList;

import junit.framework.TestCase;

import com.alexmerz.graphviz.ParseException;
import com.alexmerz.graphviz.Parser;
import com.alexmerz.graphviz.objects.Graph;


/**
 * The test.viz file contains some
 * heavy formated operations.
 * 
 * @version $Id: Parser3Test.java,v 1.5 2006/04/05 22:40:10 Alexander Exp $
 * @author Alexander
 *
 */
public class Parser3Test extends TestCase {
    
	public void testTest() throws ParseException {
        Reader in=null;
        
        try {
            InputStream inputStream = this.getClass().getClassLoader()
                    .getResourceAsStream("testfiles/test.viz");
            in = new InputStreamReader(inputStream);
            Parser p = new Parser(in);
            Boolean b = p.parse(in);
            ArrayList<Graph> al =p.getGraphs();
            for(int i=0; i<al.size();i++) {
            	System.out.println(al.get(i).toString());
            }
            assertTrue(b);                
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }        
	}
}


src/test/java/com/alexmerz/graphviz/unittest/Parser4Test.java

/***************************************************************
 *    JPGD - Java-based Parser for Graphviz Documents
 *    Copyright (C) 2006  Alexander Merz
 *
 *   This library is free software; you can redistribute it and/or
 *   modify it under the terms of the GNU Lesser General Public
 *   License as published by the Free Software Foundation; either
 *   version 2.1 of the License.
 *
 *   This library is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *   Lesser General Public License for more details.
 *
 *   You should have received a copy of the GNU Lesser General Public
 *   License along with this library; if not, write to the Free Software
 *   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */


package com.alexmerz.graphviz.unittest;

import com.alexmerz.graphviz.ParseException;
import com.alexmerz.graphviz.Parser;
import com.alexmerz.graphviz.objects.Graph;
import junit.framework.TestCase;

import java.io.*;
import java.util.ArrayList;

/**
 * This test suite uses the Graphviz examples
 * from the Gallery on the Graphviz webpage.
 * (An exception is only NodeEdge) 
 * All files are stored in the testfiles directories.
 * 
 * @version $Id: Parser2Test.java,v 1.4 2006/03/20 16:47:49 Alexander Exp $
 * @author Alexander
 *
 */
public class Parser4Test extends TestCase {

	public void testXDot() throws ParseException {
        Reader in=null;
        try {
            InputStream inputStream = this.getClass().getClassLoader()
                    .getResourceAsStream("testfiles/xdot.viz");
            in = new InputStreamReader(inputStream);
            Parser p = new Parser(in);            
            assertTrue(p.parse(in));
            ArrayList<Graph> graphs = p.getGraphs();
            graphs = null;

        } catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }               
	}
	
}


src/main/java/com/alexmerz/graphviz/objects/Edge.java
/***************************************************************
 *    JPGD - Java-based Parser for Graphviz Documents
 *    Copyright (C) 2006  Alexander Merz
 *
 *   This library is free software; you can redistribute it and/or
 *   modify it under the terms of the GNU Lesser General Public
 *   License as published by the Free Software Foundation; either
 *   version 2.1 of the License.
 *
 *   This library is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *   Lesser General Public License for more details.
 *
 *   You should have received a copy of the GNU Lesser General Public
 *   License along with this library; if not, write to the Free Software
 *   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */


package com.alexmerz.graphviz.objects;

import com.alexmerz.graphviz.Helper;

import java.util.Enumeration;
import java.util.Hashtable;

/**
 * This class represents an Edge in a Graph.
 * 
 * Each Edge has a source Node and a target Node.
 * Both nodes may contain port information.
 * 
 * The type attribute is currently a nice-to-have
 * because only directed edges are only allowed
 * in 'Digraphs', and undirected in 'Graphs'.
 * 
 * @author Alexander Merz
 * @version $Id: Edge.java,v 1.4 2006/04/05 22:39:16 Alexander Exp $
 *
 */
public class Edge {

	/**
	 * The source node object
	 */
	private PortNode source = null;
	
	/**
	 * The target node object
	 */
	private PortNode target = null;
	
	/** 
	 * The type of the edge (directed or undirected)
	 * @see Graph.EDGE_DIRECTED
	 * @see Graph.EDGE_UNDIRECTED
	 */
	private int type = 0;
	
	/**
	 * The hasttable having the attributes of this edge
	 */
	private Hashtable<String, String>attributes = new Hashtable<String, String>();
	
	/**
	 * Creates an empty edge object
	 *
	 */
	public Edge() {
		source = null;
		target = null;
		type = 0;		
	}
	
	/**
	 * Creates an Edge 
	 * 
	 * @param source the source Node
	 * @param target the target Node
	 * @param type the edge type
	 */
	public Edge(PortNode source, PortNode target, int type) {
		this.source = source;
		this.target = target;
		this.type = type;
	}
	
	/**
	 * Returns the source node of the edge
	 * @return the source node
	 */
	public PortNode getSource() {
		return source;
	}
	
	/**
	 * Sets the source node of the edge
	 * @param source the source node
	 */
	public void setSource(PortNode source) {
		this.source = source;
	}
	
	/**
	 * Returns the target node of the edge
	 * @return the target node
	 */
	public PortNode getTarget() {
		return target;
	}
	
	/**
	 * Sets the target node of the edge
	 * @param target the target Node
	 */
	public void setTarget(PortNode target) {
		this.target = target;
	}
	
	/**
	 * Returns the type of the edge.
	 * @return the type of the edge
	 * @see com.alexmerz.graphviz.objects.Graph#UNDIRECTED
	 * @see com.alexmerz.graphviz.objects.Graph#DIRECTED
	 */
	public int getType() {
		return type;
	}
	
	/**
	 * Sets the type of the edge
	 * @param type the type of the edge
	 * @see com.alexmerz.graphviz.objects.Graph#UNDIRECTED
	 * @see com.alexmerz.graphviz.objects.Graph#DIRECTED	 
	 */
	public void setType(int type) {
		this.type = type;
	}
	
	/**
	 * Returns the String representation of the edge
	 * @return the string representation
	 */
	public String toString() {
		StringBuffer r = new StringBuffer();
		if(!source.getNode().getId().getId().equals("")) {
			r.append(source.getNode().getId().getId());
		} else if(!source.getNode().getId().getLabel().equals("")) {
			r.append("\"");
			r.append(source.getNode().getId().getLabel());
			r.append("\"");
		}
		if(!source.getPort().equals("")) {
			r.append(":\"");
			r.append(source.getPort());
			r.append("\"");
		}
		if(Graph.DIRECTED == getType()) {
			r.append(" -> ");
		} else {
			r.append(" -- ");
		}
		if(!target.getNode().getId().getId().equals("")) {
			r.append(target.getNode().getId().getId());
		} else if(!target.getNode().getId().getLabel().equals("")) {
			r.append("\"");
			r.append(target.getNode().getId().getLabel());
			r.append("\"");
		}
		if(!target.getPort().equals("")) {
			r.append(":\"");
			r.append(target.getPort());
			r.append("\"");
		}		
		if(attributes.size()>0) {
			r.append(" [");
			Enumeration<String> e = attributes.keys();
			while(e.hasMoreElements()) {
				String k = e.nextElement();
				r.append(k);
				{
					r.append("=");
					String attr = attributes.get(k);
					if(!Helper.needsQuote(attr)) {
						r.append(attr);
					} else {
						r.append("\"");
						r.append(attr);
						r.append("\"");					
					}
				}
				r.append(", ");
			}
			r.delete(r.length()-2, r.length());
			r.append("]");
		}		
		r.append(";\n");		
		return r.toString();
	}

	/**
	 * Returns the value of an edge attribute
	 * @param key the name of the attribute
	 * @return the value of the attribute
	 */
	public String getAttribute(String key) {
		return attributes.get(key);
	}

	/**
	 * Sets the value of an attribute
	 * @param key the name of the attribute
	 * @param value the value of the attribute
	 */
	public void setAttribute(String key, String value) {	
		this.attributes.put(key, value);
	}
	
	/**
	 * Returns all attributes of this edge
	 * @return the attributes
	 */
	public Hashtable<String,String> getAttributes() {
		return this.attributes;
	}
	
}


src/main/java/com/alexmerz/graphviz/objects/Node.java
/***************************************************************
 *    JPGD - Java-based Parser for Graphviz Documents
 *    Copyright (C) 2006  Alexander Merz
 *
 *   This library is free software; you can redistribute it and/or
 *   modify it under the terms of the GNU Lesser General Public
 *   License as published by the Free Software Foundation; either
 *   version 2.1 of the License.
 *
 *   This library is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *   Lesser General Public License for more details.
 *
 *   You should have received a copy of the GNU Lesser General Public
 *   License along with this library; if not, write to the Free Software
 *   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

package com.alexmerz.graphviz.objects;

import com.alexmerz.graphviz.Helper;

import java.util.Enumeration;
import java.util.Hashtable;

/**
 * This class represents a node in a graph.
 * 
 * @author Alexander Merz
 * @version $Id: Node.java,v 1.4 2006/04/05 22:39:30 Alexander Exp $
 *
 */
public class Node {

	/**
	 * The id object
	 */
	private Id id = null;
	
	/**
	 * Holds the attributes of the node
	 */
	private Hashtable<String, String> attr = new Hashtable<String, String>(); 
	
	/**
	 * If true this node represents a subgraph
	 */
	private boolean isSubgraph = false;
	
	/**
	 * Returns the id object for the node
	 * @return the Id object
	 */
	public Id getId() {
		return id;
	}
	
	/**
	 * Sets the Id object for this node
	 * @param id
	 */
	public void setId(Id id) {
		this.id = id;
	}

	/**
	 * Returns a string representation of this node
	 * If the node is a subgraph, an empty string will be returned
	 * @return the string representation
	 */
	public String toString() {
		if(isSubgraph()) {
			return "";
		}
		StringBuffer r = new StringBuffer("");
		if(!id.getId().equals("")) {
			r.append(id.getId());
		} else if(!id.getLabel().equals("")) {
			r.append("\"");
			r.append(id.getLabel());
			r.append("\"");
		}		
		if(attr.size()>0) {
			r.append(" [");
			Enumeration<String> e = attr.keys();
			while(e.hasMoreElements()) {
				String k = e.nextElement();
				r.append(k);
				{
					r.append("=");
					if(!Helper.needsQuote(attr.get(k))) {
						r.append(attr.get(k));
					} else {
						r.append("\"");
						r.append(attr.get(k));
						r.append("\"");					
					}
				}
				r.append(", ");
			}
			r.delete(r.length()-2, r.length());			
			r.append("]");
		}
		r.append(";\n");
		return r.toString();
	}

	/**
	 * Returns the attribute of the node
	 * @param key the name of the attribute
	 * @return the value of the attribute
	 */
	public String getAttribute(String key) {
		return attr.get(key);
	}

	/**
	 * Sets the attribute of the node
	 * @param key the name of the attribute
	 * @param value the value of the attribute
	 */
	public void setAttribute(String key, String value) {
		this.attr.put(key, value);
	}

	/**
	 * Returns true, if the node object represents 
	 * a subgraph.
	 * @return true or false
	 */
	public boolean isSubgraph() {
		return isSubgraph;
	}

	/**
	 * Sets, if the node represents a subgraph
	 * @param isSubgraph true if subgraph is used in
	 *        edge operation
	 */
	public void representsSubgraph(boolean isSubgraph) {
		this.isSubgraph = isSubgraph;
	}
	
	/**
	 * Returns all attributes of the edge.
	 * @return the edge attributes.
	 */
	public Hashtable<String,String> getAttributes() {
		return this.attr;
	}
}


src/main/java/com/alexmerz/graphviz/objects/Graph.java
/***************************************************************
 *    JPGD - Java-based Parser for Graphviz Documents
 *    Copyright (C) 2006, 2015  Alexander Merz, Braxton Fitts
 *
 *   This library is free software; you can redistribute it and/or
 *   modify it under the terms of the GNU Lesser General Public
 *   License as published by the Free Software Foundation; either
 *   version 2.1 of the License.
 *
 *   This library is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *   Lesser General Public License for more details.
 *
 *   You should have received a copy of the GNU Lesser General Public
 *   License along with this library; if not, write to the Free Software
 *   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

package com.alexmerz.graphviz.objects;

import com.alexmerz.graphviz.Helper;

import java.util.ArrayList;
import java.util.Enumeration;
import java.util.Hashtable;

/**
 * This class represents a graph.
 * 
 * A Graph contains Nodes, Edges and Subgraphs
 * 
 * @author Alexander
 * @author Braxton
 * @version $Id: Graph.java,v 1.5 2006/04/05 22:39:23 Alexander Exp $
 *
 */
public class Graph {

	/* Constants */
	/**
	 * Constant for an undirected graph
	 */
	public static final int UNDIRECTED = 1;
	
	/**
	 * Constant for an directed graph
	 */
	public static final int DIRECTED = 2;
	
	/**
	 * Identifier object for the graph
	 */
	private Id id=null;
	
	/**
	 * Holds all Node objects of this graph
	 */
	private ArrayList<Node> nodes = new ArrayList<Node>();
	
	/**
	 * Holds all Edge objects of this graph
	 */
	private ArrayList<Edge> edges = new ArrayList<Edge>();

	/**
	 * Holds all Subgraphs
	 */
	private ArrayList<Graph> graphs = new ArrayList<Graph>();	
	
	/**
	 * The type of this graph
	 * @see DIRECTED
	 * @see UNDIRECTED 
	 */
	private int type = 0;
	
	/**
	 * Wether the graph is strict or not
	 */
	private boolean strictgraph = false;
	
	/**
	 * Holds the graph attributes
	 */
	private Hashtable<String,String> attributes = new Hashtable<String, String>();
	
	/**
	 * Holds generic attributes for all edges
	 */
	private Hashtable<String,String> genericEdgeAttributes = new Hashtable<String, String>();
	
	/**
	 * Holds generic attributes for all nodes 
	 */
	private Hashtable<String,String> genericNodeAttributes = new Hashtable<String, String>();

	/**
	 * Holds generic attributes for graphs 
	 */
	private Hashtable<String,String> genericGraphAttributes = new Hashtable<String, String>();	
	
	/**
	 * Sets a generic attribute for all edges of this graph.
	 * This attributes are NOT additionally stored in the Edge objects. 
	 * 
	 * @param key the attribute name
	 * @param value the attribute value
	 */
	public void addGenericEdgeAttribute(String key, String value) {
		this.genericEdgeAttributes.put(key, value);
	}
	
	/**
	 * Returns a generic attribute of an edge
	 * @param key the attribute name
	 * @return the value of the attribute or null, if the attribute does not exist.
	 */
	public String getGenericEdgeAttribute(String key) {
		return (String)this.genericEdgeAttributes.get(key);
	}

	/**
	 * Sets a generic attribute for clusters of this graph.
	 * This attributes are NOT additionally stored in the Edge objects. 
	 * 
	 * @param key the attribute name
	 * @param value the attribute value
	 */
	public void addGenericGraphAttribute(String key, String value) {
		this.genericGraphAttributes.put(key, value);
	}
	
	/**
	 * Returns a generic attribute of the clusters in the graph
	 * @param key the attribute name
	 * @return the value of the attribute or null, if the attribute does not exist.
	 */
	public String getGenericGraphAttribute(String key) {
		return (String)this.genericGraphAttributes.get(key);
	}
	
	
	/**
	 * Sets a generic attribute for all nodes of this graph
	 * This attributes are NOT additionally stored in the Node objects.
	 * @param key the attribute name
	 * @param value the attribute value
	 */
	public void addGenericNodeAttribute(String key, String value) {
		this.genericNodeAttributes.put(key, value);
	}
	
	/**
	 * Returns a generic attribute of a node
	 * @param key the attribute name
	 * @return the value of the attribute or null, if the attribute does not exist.
	 */
	public String getGenericNodeAttribute(String key) {
		return (String)this.genericNodeAttributes.get(key);
	}	
	
	/**
	 * Adds a graph attribute.
	 * This attributes are NOT inherited to the attributes of sub graphs.
	 * 
	 * @param key the name of the attribute
	 * @param value the value of the attribute
	 */
	public void addAttribute(String key, String value) {
		attributes.put(key, value);
	}
	
	/**
	 * Returns an attribute of the Graph or null if not found.
	 * @param key the name of the attribute
	 * @return the value of the attribute
	 */
	public String getAttribute(String key) {
		return (String) attributes.get(key);
	}	
	
	/**
	 * Adds a Node object to the graph
	 * @param n the Node Object to add
	 */
	public void addNode(Node n) {
		nodes.add(n);
	}
	
	
	/**
	 * Adds an Edge object to the graph
	 * @param e the Edge object to add
	 */
	public void addEdge(Edge e) {
		edges.add(e);
	}
	
	/**
	 * Returns the Id object to the Graph
	 * @return the Id object
	 */
	public Id getId() {
		return id;
	}
	
	/**
	 * Sets the Id object for this graph
	 * @param id
	 */
	public void setId(Id id) {
		this.id = id;
	}

	/**
	 * Returns the type of the graph
	 * 
	 * @return the type of the graph
	 * @see com.alexmerz.graphviz.objects.Graph#DIRECTED
	 * @see com.alexmerz.graphviz.objects.Graph#UNDIRECTED
	 */
	public int getType() {
		return type;
	}
	
	/**
	 * Sets the type of the graph
	 * @param type the type of the graph
	 * @see com.alexmerz.graphviz.objects.Graph#DIRECTED
	 * @see com.alexmerz.graphviz.objects.Graph#UNDIRECTED
	 */
	public void setType(int type) {
		if(graphs.size()!=0) {			
			for(int i=0; i<graphs.size(); i++) {
				graphs.get(i).setType(type);				
			}
		}
		this.type = type;
	}
	
	/**
	 * Returns wether the graph is strict or not
	 * @return the static state
	 */
	public boolean isStrict() {
		return strictgraph;
	}
	
	/**
	 * Sets if the graph is strict or not.
	 * @param isStrict
	 */
	public void setStrict(boolean isStrict) {
		this.strictgraph = isStrict;
	}

	/**
	 * Returns a String representation of the graph
	 * @return the string representation
	 */
	@Override
	public String toString() {
		return toString(null);
	}

	public String toString(String head) {
		StringBuffer r = new StringBuffer();
		if(isStrict()) {
			r.append("strict ");
		}
		if (head != null) {
			r.append(head).append(" ");
		} else {
			if (DIRECTED == getType()) {
				r.append("digraph ");
			} else {
				r.append("graph ");
			}
		}
		if(!id.getId().equals("")) {
			r.append(id.getId());
		} else if(!id.getLabel().equals("")) {
			r.append(id.getLabel());
		}
		r.append(" {\n");

		if(attributes.size()>0) { 		
			Enumeration<String> e = attributes.keys();
			String k;
			while(e.hasMoreElements()) {
				k = e.nextElement();
				r.append(k);
				String attr = attributes.get(k);
				{
					r.append("=");
					if(!Helper.needsQuote(attr)) {
						r.append(attr);
					} else {
						r.append("\"");
						r.append(attr);
						r.append("\"");					
					}					
				}
				r.append(";\n");				
			}			
		}		
		
		if(genericNodeAttributes.size()>0) { 		
			Enumeration<String> e = genericNodeAttributes.keys();
			String k;
			r.append("node [");
			while(e.hasMoreElements()) {
				k = e.nextElement();
				r.append(k);
				String attr = genericNodeAttributes.get(k);
				{
					r.append("=");
					if(!Helper.needsQuote(attr)) {
						r.append(attr);
					} else {
						r.append("\"");
						r.append(attr);
						r.append("\"");					
					}			
				}
				r.append(", ");
			}
			r.delete(r.length()-2, r.length());			
			r.append("];\n");
		}
		
		if(genericEdgeAttributes.size()>0) { 		
			Enumeration<String> e = genericEdgeAttributes.keys();
			String k;
			r.append("edge [");
			while(e.hasMoreElements()) {
				k = e.nextElement();
				r.append(k);
				String attr = genericEdgeAttributes.get(k);
				{
					r.append("=");
					if(!Helper.needsQuote(attr)) {
						r.append(attr);
					} else {
						r.append("\"");
						r.append(attr);
						r.append("\"");					
					}			
				}
				r.append(", ");				
			}
			r.delete(r.length()-2, r.length());			
			r.append("];\n");
		}		

		if(genericGraphAttributes.size()>0) { 		
			Enumeration<String> e = genericGraphAttributes.keys();
			String k;
			r.append("graph [");
			while(e.hasMoreElements()) {
				k = e.nextElement();
				r.append(k);
				String attr = genericGraphAttributes.get(k);
				{
					r.append("=");
					if(!Helper.needsQuote(attr)) {
						r.append(attr);
					} else {
						r.append("\"");
						r.append(attr);
						r.append("\"");					
					}
				}
				r.append(", ");
			}
			r.delete(r.length()-2, r.length());			
			r.append("];\n");
		}		
		if(nodes.size()>0) {
			for(int i=0; i<nodes.size(); i++) {
				r.append(nodes.get(i).toString());
			}
		}
		if(edges.size()>0) {
			for(int i=0; i<edges.size(); i++) {
				r.append(edges.get(i).toString());
			}
		}
		if(graphs.size()>0) {
			for(int i=0; i<graphs.size(); i++) {
				Graph graph = graphs.get(i);
				String gHead;
				if (graph.getId().getId().startsWith("cluster") || graph.getId().getId().startsWith("rankAligner")) {
					gHead = "subgraph";
				} else {
					gHead = null;
				}
				r.append(graph.toString(gHead));
				r.append("\n");
			}
		}		
		r.append(" }");
		return r.toString();
/*		
		return "Graph: " + ((id!=null)?id.toString():"") + ", "+
				"static="+Boolean.toString(strictgraph)+", "+
				((type==DIRECTED)?"directed":"undirected")+
				", Attributes: "+attributes.toString()+
				", Nodes: "+nodes.toString()+
				", Edges: "+edges.toString()+
				", Subgraphs: "+graphs.toString();							
		*/
	}

	/**
	 * Returns a list of all sub graphs. 
	 * @return the subgraphs
	 */
	public ArrayList<Graph> getSubgraphs() {
		return graphs;
	}

	/** 
	 * Adds a sub graph to a graph
	 * @param graph
	 */
	public void addSubgraph(Graph graph) {
		this.graphs.add(graph);
	}
	
	/**
	 * Tries to find a node of a Graph depending
	 * on the given Id object.
	 * @param id the id object to identify the node
	 * @return the node or nulll if not found
	 */
	public Node findNode(Id id) {
		Node n = null;
		Id nid = null;
				
		for(int i=0; i<nodes.size(); i++) {
			n = nodes.get(i);
			nid = n.getId();
			if(nid.isEqual(id)) {
				return n;
			}
		}
		Graph g;
		for(int i=0; i<graphs.size(); i++) {
			g = graphs.get(i);
			n = g.findNode(id);
			if(n!=null) {
				return n;
			}
		}
		return null;
	}

	/**
	 * Returns all Nodes of the graph.
	 * 
	 * If the graph contains subgraphs, you can use
	 * the thisOnly parameter to decide if nodes defined in subgraphs
	 * should be added to the list also.
	 * 
	 * @param thisOnly if true, also include nodes in subgraphs also, else exclude them
	 * @return the nodes of the graph
	 */
	public ArrayList<Node> getNodes(boolean thisOnly) {
		if(thisOnly) {
			return this.nodes;
		} else {
			ArrayList<Node> n = new ArrayList<Node>(this.nodes);
			Graph g = null;
			for(int i=0;i<this.graphs.size();i++) {
				g = graphs.get(i);
				n.addAll(g.getNodes(false));
			}
			return n;
		}
	}
	
	/**
	 * Returns all attributes of the graph.
	 * @return the attributes
	 */
	public Hashtable<String, String> getAttributes() {
		return this.attributes;
	}
	
	/**
	 * Returns all edges of this graph.
	 * @return a list of Edge objects
	 */
	public ArrayList<Edge> getEdges() {
		return this.edges;
	}
}
